openmls/src/group/mls_group/staged_commit.rs

```diff
@@ -290,13 +289,39 @@ impl MlsGroup {
     pub fn merge_commit(&mut self, staged_commit: StagedCommit) -> Result<(), MlsGroupError> {
         if let Some(state) = staged_commit.state {
             self.group_context = state.group_context;
-            self.epoch_secrets = state.epoch_secrets;
+            self.group_epoch_secrets = state.group_epoch_secrets;
+            self.message_secrets = state.message_secrets;
             self.interim_transcript_hash = state.interim_transcript_hash;
-            self.secret_tree = state.secret_tree;
             self.tree.merge_diff(state.staged_diff)?;
         };
         Ok(())
     }
+
+    /// Merges a [StagedCommit] into the group state and optionally return a [`SecretTree`]
+    /// from the previous epoch. The secret tree is returned if the Commit does not contain a self removal.
+    ///
+    /// This function should not fail and only returns a [`Result`], because it
+    /// might throw a `LibraryError`.
+    pub fn merge_commit_take_message_secrets(
+        &mut self,
+        staged_commit: StagedCommit,
+    ) -> Result<Option<MessageSecrets>, MlsGroupError> {
```

Comment:
If there's no point in failing it can as well just return an `Option`.

Reply:
No, because as the comment says, these are different cases. If we really throw a `LibraryError` but only return a `None`, applications could get confused if they expected a `Some(...)`. We'd make it their responsibility to detect it was a library error. I'd rather throw the error early on and give applications something they can easily react to.

Reply:
Sure if you really want to. It's just not very idiomatic. Rust callers would expect an `Option` here as result.

Reply:
I agree it doesn't look great. I'm happy to change it for something else, as long as it doesn't introduce dangerous behaviour. Maybe we can get rid of the library errors at some point.